tests/unit/test_cmac.cpp

```diff
@@ -0,0 +1,267 @@
+                    "f69f2445df4f9b17ad2b417be66c3710",
+                "51f0bebf7e3b9d92fc49741779363cfe",
+            },
+
+            // Test data for AES-256.
+            {
+                openssl::CmacCipherTypes::AES_CBC_256,
+                "2b7e151628aed2a6abf7158809cf4f3c"
+                    "deadbeef38317deafbeef3831700cafe",
+                "",
+                "9922123e226ee972e5fd501e45cae51d",
+            },
+            {
+                openssl::CmacCipherTypes::AES_CBC_256,
+                "2b7e151628aed2a6abf7158809cf4f3c"
+                    "deadbeef38317deafbeef3831700cafe",
+                "6bc1bee22e409f96e93d7e117393172a",
+                "9414636aa65c795a59e8813e2fbb588a",
+            },
+            {
+                openssl::CmacCipherTypes::AES_CBC_256,
+                "2b7e151628aed2a6abf7158809cf4f3c"
+                    "deadbeef38317deafbeef3831700cafe",
+                "6bc1bee22e409f96e93d7e117393172a"
+                    "ae2d8a571e03ac9c9eb76fac45af8e51"
+                    "30c81c46a35ce411",
+                "bf20027ad24e648f88d7cf0d2eb03f93",
+            },
+            {
+                openssl::CmacCipherTypes::AES_CBC_256,
+                "2b7e151628aed2a6abf7158809cf4f3c"
+                    "deadbeef38317deafbeef3831700cafe",
+                "6bc1bee22e409f96e93d7e117393172a"
+                    "ae2d8a571e03ac9c9eb76fac45af8e51"
+                    "30c81c46a35ce411e5fbc1191a0a52ef"
+                    "f69f2445df4f9b17ad2b417be66c3710",
+                "3598968688288d17a2f5544cef0651ea",
+            },
+        }
+    };
+
+    return testdata;
+}
+
+class CalculateCmacValues : public testing::TestWithParam<Testdata> {};
+
+INSTANTIATE_TEST_CASE_P(
+        cmac,
+        CalculateCmacValues,
+        testing::ValuesIn(prepareTestdataForCmacTests()));
+
+TEST_P(CalculateCmacValues, returnsCorrectCmac)
+{
+    auto testdata = GetParam();
+
+    auto cmacCalculator = CMAC(testdata.cipherType, utility::fromHex(testdata.key));
+    cmacCalculator.update(utility::fromHex(testdata.message));
+    auto actualCmac = cmacCalculator.finish();
+
+    ASSERT_THAT(utility::toHex(actualCmac), Eq(testdata.expectedCmac));
+}
+
+TEST_P(CalculateCmacValues, verifyIsSuccessfulForCorrectCmac)
+{
+    auto testdata = GetParam();
+
+    auto cmacCalculator = CMAC(testdata.cipherType, utility::fromHex(testdata.key));
+    cmacCalculator.update(utility::fromHex(testdata.message));
+    cmacCalculator.finish();
+
+    auto expectedCmac = utility::fromHex(testdata.expectedCmac);
+    cmacCalculator.verify(expectedCmac);
+}
+
+TEST_P(CalculateCmacValues, verifyFailsForIncorrectCmac)
+{
+    auto testdata = GetParam();
+
+    auto cmacCalculator = CMAC(testdata.cipherType, utility::fromHex(testdata.key));
+    cmacCalculator.update(utility::fromHex(testdata.message));
+    cmacCalculator.finish();
+
+    auto incorrectCmac = utility::fromHex(testdata.expectedCmac);
+    incorrectCmac[0]++;
```

Comment:
Strictly speaking this opens the door to undefined behavior. Maybe we want to be a good example and do incorrectCmac[0] ^= 0x1;

Reply:
I appreciate that.
Just to sort it out correctly for me: This particular case is well defined (unsigned integral type) but to avoid potentially problems, such a pattern should be avoided if possible, right?

Reply:
A true, I didn't check that its unsigned. So you're right. Maybe we want to change this anyways?

Reply:
Sure.

Reply:
Done.